Replace hex code point enumerations with bitfield syntax (#1774)
moq-wg/moq-transport: draft-ietf-moq-transport.md

~~~
OBJECT_DATAGRAM {

>   Type (vi64) = 0b00X0XXXX,

Reviewer: This is strange because it makes it look like it's an 8 bit value, when in fact it's a varint that currently fits in 8 bits.  We could rename it "Type Flags (vi64)" and omit the bit pattern here, it's explained below.

~~~
{: #object-datagram-format title="MOQT OBJECT_DATAGRAM"}

The Type field in the OBJECT_DATAGRAM is a single-byte vi64 with the following
bit layout:

~~~

>   7   6   5   4   3   2   1   0

Reviewer: Bit 7 is interesting because of the varint encoding intersection.  We probably need to say something about it.

+---+---+---+---+---+---+---+---+
| 0 | 0 | S | 0 | D | Z | E | P |
+---+---+---+---+---+---+---+---+

P = PROPERTIES (0x01)
E = END_OF_GROUP (0x02)
Z = ZERO_OBJECT_ID (0x04)
D = DEFAULT_PRIORITY (0x08)
S = STATUS (0x20)
~~~

> Not all Type values in this range are valid. The four low-order bits and bit 5

Reviewer: ```suggestion
The four low-order bits and bit 5
```

Reviewer: It should be a protocol violation if a 0 bit is set, or any bits higher than bit 6 are set.

~~~
{: #object-datagram-format title="MOQT OBJECT_DATAGRAM"}

> The Type field in the OBJECT_DATAGRAM is a single-byte vi64 with the following

Reviewer 2: It technically does not have to be one byte, it merely is small enough to always fit into one.

Author: I feel it makes more sense to keep this flags to be 1 byte ? Do we expect to be have more than 1 byte of values ?

bit layout:

~~~
  7   6   5   4   3   2   1   0
+---+---+---+---+---+---+---+---+
| 0 | 0 | S | 0 | D | Z | E | P |
+---+---+---+---+---+---+---+---+

P = PROPERTIES (0x01)
E = END_OF_GROUP (0x02)
Z = ZERO_OBJECT_ID (0x04)
D = DEFAULT_PRIORITY (0x08)
S = STATUS (0x20)

> ~~~

Reviewer 2: Should this use the QUIC syntax used everywhere else in the document?

~~~
SUBGROUP_HEADER {

>   Type (vi64) = 0b0XX1XXXX,

Reviewer: Same comments as above

~~~
{: #object-datagram-format title="MOQT OBJECT_DATAGRAM"}

The Type Flags field in the OBJECT_DATAGRAM is a variable-length integer that
encodes a set of flags. All valid values fit in a single-byte encoding (values
less than 128). If a received value has bit 4 set, or has a

> value of 64 or greater, the endpoint MUST close the session with a

Reviewer: ```suggestion
less than 128). If a received value has bit 4 set, or has a
bit set whose meaning is not specified, the endpoint MUST close the session with a
```

> `PROTOCOL_VIOLATION`.

Reviewer: ```suggestion
The Type Flags field in the OBJECT_DATAGRAM is a variable-length integer that
encodes a set of flags. All values defined in this specification fit in a
single-byte encoding (values less than 128). If a received value has bit 4 set,
or has a bit set whose meaning is not specified, the endpoint MUST close the
session with a `PROTOCOL_VIOLATION`.
```

The four low-order bits and bit 5

> of the Type Flags field determine which fields are present in the datagram:

Reviewer: ```suggestion
The four low-order bits and bit 5 of the Type Flags field determine which fields
are present in the datagram:
```

~~~
{: #object-datagram-format title="MOQT OBJECT_DATAGRAM"}

The Type Flags field in the OBJECT_DATAGRAM is a variable-length integer that

> encodes a set of flags. All valid values fit in a single-byte encoding (values

Reviewer: ```suggestion
encodes a set of flags. All values defined in this specification fit in a single-byte encoding (values
```

less than 64). Bits 7 and 6 are used by the variable-length integer encoding;

> the flags occupy bits 5 through 0. If a received value has bit 4 set, or has a

Reviewer 3: ```suggestion
less than 128). If a received value has bit 4 set, or has a
```

  the Object Payload; the entirety of the transport datagram following the
  Object header contains the payload.

The following Type Flags values are invalid. If an endpoint receives a datagram
with any of these values, it MUST close the session with a `PROTOCOL_VIOLATION`:

* Values with both the STATUS bit (0x20) and END_OF_GROUP bit (0x02) set: 0x22,

>   0x23, 0x26, 0x27, 0x2A, 0x2B, 0x2E, 0x2F. An object status message cannot signal

Reviewer: ```suggestion
* Values with both the STATUS bit (0x20) and END_OF_GROUP bit (0x02) set.
```

>   end of group.

Reviewer: ```suggestion
```

* Values with bit 4 (0x10) set. This bit is reserved and MUST be zero.

* Values of 64 or greater (i.e., any value that requires more than a single-byte

>   variable-length integer encoding).

Reviewer: ```suggestion
* Values with a bit set whose meaning is not specified.
```

  the Object Payload; the entirety of the transport datagram following the
  Object header contains the payload.

The following Type Flags values are invalid. If an endpoint receives a datagram
with any of these values, it MUST close the session with a `PROTOCOL_VIOLATION`:

* Values with both the STATUS bit (0x20) and END_OF_GROUP bit (0x02) set: 0x22,
  0x23, 0x26, 0x27, 0x2A, 0x2B, 0x2E, 0x2F. An object status message cannot signal
  end of group.

* Values with bit 4 (0x10) set. This bit is reserved and MUST be zero.

* Values of 64 or greater (i.e., any value that requires more than a single-byte
  variable-length integer encoding). Bits 7 and 6 are used by the variable-length

>   integer encoding and MUST be zero for OBJECT_DATAGRAM.

Reviewer 3: ```suggestion
  variable-length integer encoding).
```

All Objects received on a stream opened with `SUBGROUP_HEADER` have an
`Object Forwarding Preference` = `Subgroup`.

The Type Flags field in the SUBGROUP_HEADER is a variable-length integer that
encodes a set of flags. All values defined in this specification fit in a single-byte encoding (values

> less than 128).

Reviewer: ```suggestion
The Type Flags field in the SUBGROUP_HEADER is a variable-length integer that
encodes a set of flags. All values defined in this specification fit in a
single-byte encoding (values less than 128).
```

All Objects received on a stream opened with `SUBGROUP_HEADER` have an
`Object Forwarding Preference` = `Subgroup`.

The Type Flags field in the SUBGROUP_HEADER is a variable-length integer that
encodes a set of flags. All valid values fit in a single-byte encoding (values
less than 128). Bits 7 and 6 are used by the variable-length integer encoding;
however, bit 6 also carries the FIRST_OBJECT flag. Since a value with bit 6 set
requires a two-byte varint encoding, implementations MUST use the two-byte form

> when FIRST_OBJECT is set.

Reviewer: ```suggestion
The Type Flags field in the SUBGROUP_HEADER is a variable-length integer that
encodes a set of flags. All values defined in this specification fit in a single-byte encoding (values
less than 128).
```

All Objects received on a stream opened with `SUBGROUP_HEADER` have an
`Object Forwarding Preference` = `Subgroup`.

The Type Flags field in the SUBGROUP_HEADER is a variable-length integer that
encodes a set of flags. All valid values fit in a single-byte encoding (values
less than 128). Bits 7 and 6 are used by the variable-length integer encoding;
however, bit 6 also carries the FIRST_OBJECT flag. Since a value with bit 6 set
requires a two-byte varint encoding, implementations MUST use the two-byte form
when FIRST_OBJECT is set. If a received value has bit 7 set (i.e., uses a
four-byte or eight-byte varint encoding), the endpoint MUST close the session

> with a `PROTOCOL_VIOLATION`.

Reviewer: ```suggestion
encodes a set of flags. All values defined in this specification fit in a single-byte encoding (values
less than 128). 
```

All Objects received on a stream opened with `SUBGROUP_HEADER` have an
`Object Forwarding Preference` = `Subgroup`.

The Type Flags field in the SUBGROUP_HEADER is a variable-length integer that
encodes a set of flags. All valid values fit in a single-byte encoding (values
less than 128). Bits 7 and 6 are used by the variable-length integer encoding;
however, bit 6 also carries the FIRST_OBJECT flag. Since a value with bit 6 set
requires a two-byte varint encoding, implementations MUST use the two-byte form
when FIRST_OBJECT is set. If a received value has bit 7 set (i.e., uses a
four-byte or eight-byte varint encoding), the endpoint MUST close the session

> with a `PROTOCOL_VIOLATION`.

Reviewer 3: ```suggestion
when FIRST_OBJECT is set.
```

* The **FIRST_OBJECT** bit (0x40) indicates that the first object in this
subgroup stream is the first object published in the subgroup by the original publisher.

The following Type Flags values are invalid. If an endpoint receives a stream
header with any of these values, it MUST close the session with a
`PROTOCOL_VIOLATION`:

* Values with SUBGROUP_ID_MODE set to 0b11: 0x16, 0x17, 0x1E, 0x1F, 0x36,

>   0x37, 0x3E, 0x3F, 0x56, 0x57, 0x5E, 0x5F, 0x76, 0x77, 0x7E, 0x7F. This mode

Reviewer: ```suggestion
* Values with SUBGROUP_ID_MODE set to 0b11. This mode
```

  is reserved for future use.

* Values where bit 4 is not set. Bit 4 MUST be 1 for SUBGROUP_HEADER.

* Values of 128 or greater (i.e., any value that requires more than a two-byte
  variable-length integer encoding). Bit 7 is used by the variable-length

>   integer encoding and MUST be zero for SUBGROUP_HEADER.

Reviewer: ```suggestion
* Values with a bit set whose meaning is not specified.
```

* The **FIRST_OBJECT** bit (0x40) indicates that the first object in this
subgroup stream is the first object published in the subgroup by the original publisher.

The following Type Flags values are invalid. If an endpoint receives a stream
header with any of these values, it MUST close the session with a
`PROTOCOL_VIOLATION`:

* Values with SUBGROUP_ID_MODE set to 0b11: 0x16, 0x17, 0x1E, 0x1F, 0x36,
  0x37, 0x3E, 0x3F, 0x56, 0x57, 0x5E, 0x5F, 0x76, 0x77, 0x7E, 0x7F. This mode
  is reserved for future use.

* Values where bit 4 is not set. Bit 4 MUST be 1 for SUBGROUP_HEADER.

* Values of 128 or greater (i.e., any value that requires more than a two-byte
  variable-length integer encoding). Bit 7 is used by the variable-length

>   integer encoding and MUST be zero for SUBGROUP_HEADER.

Reviewer 3: ```suggestion
  variable-length integer encoding).
```
